jsonschema/reference.go

```diff
@@ -0,0 +1,161 @@
+package jsonschema
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/franela/goreq"
+)
+
+// jsonRef represents a JSON Reference source and targetRef
+type jsonRef struct {
+	Source []string
+	Target string
+}
+
+// dereference parses JSON string and replaces all $ref with the referenced data.
+func dereference(schemaPath string, input []byte, isTop bool) ([]byte, error) {
+	if !strings.Contains(string(input), "$ref") {
+		return input, nil
+	}
+
+	var data interface{}
+	json.Unmarshal([]byte(input), &data)
+	refs, err := walkInterface(data, []string{}, []jsonRef{})
+	if err != nil {
+		return input, fmt.Errorf("unable to walk interface %s: %v", schemaPath, err)
+	}
+
+	for _, ref := range refs {
+		top := data
+		for i, item := range ref.Source {
+			if isTop && (ref.Source[0] == "allOf" || ref.Source[0] == "oneOf") {
+				continue // do not dereference top-level allOr and oneOf
+			}
+			if i < len(ref.Source)-1 { // iterate
+				if intKey, err := strconv.Atoi(item); err == nil {
+					top = top.([]interface{})[intKey]
+				} else {
+					top = top.(map[string]interface{})[item]
+				}
+			} else { // set reference
+				targetRef, err := buildReference(schemaPath, data, ref.Target)
+				if err != nil {
+					return input, fmt.Errorf("unable to build reference from %s: %v", ref.Target, err)
+				}
+				targetKeys := reflect.ValueOf(targetRef).MapKeys()
+				if len(targetKeys) > 1 {
+					// assuming integer item is slice[index] instead of map[string]
+					if intKey, err := strconv.Atoi(item); err == nil {
+						top.([]interface{})[intKey] = targetRef
+					} else {
```

Comment:
I would be really curious about running gocyclo on this to find the cyclomatic complexity. I'm out of time to think harder about how to simplify it but it looks really complex on first glance.

Reply:
```
$ gocyclo -top 10 ./jsonschema/
18 jsonschema TestWalkJSONSchema jsonschema/walk_test.go:20:1
14 jsonschema dereference jsonschema/reference.go:23:1
14 jsonschema TestWalkJSONSchemaRaw jsonschema/walk_test.go:292:1
12 jsonschema SchemaFromFile jsonschema/schema.go:49:1
11 jsonschema TestSchemaFromFile jsonschema/schema_test.go:82:1
11 jsonschema TestDereference jsonschema/reference_test.go:14:1
11 jsonschema walkInstance jsonschema/walk.go:52:1
10 jsonschema buildReference jsonschema/reference.go:113:1
10 jsonschema walkInterface jsonschema/reference.go:80:1
6 jsonschema TestValidate jsonschema/validator_test.go:8:1
```

Reply:
Thanks for running that, looks like I was right, other than some tests that I wrote it is the most complicated. More motivation to circle back and try and clean it all up later.